go/runtime/proxy.go

```diff
@@ -2,269 +2,174 @@ package runtime
-	}
-	if proxyOpenResp.OpenResponse == nil {
-		return fmt.Errorf("internal protocol error, expected OpenResponse from container")
-	}
-	if proxyOpenResp.OpenResponse.Status != pf.TaskNetworkProxyResponse_OK {
-		logrus.WithFields(logrus.Fields{
-			"status":   proxyOpenResp.OpenResponse.Status,
-			"shardID":  open.ShardId,
-			"portName": open.TargetPort,
-		}).Warn("connector-init returned !OK")
-		openResp.Status = proxyOpenResp.OpenResponse.Status
-		return streaming.Send(&pf.TaskNetworkProxyResponse{
-			OpenResponse: openResp,
-		})
-	}
-
-	if err = streaming.Send(&pf.TaskNetworkProxyResponse{OpenResponse: openResp}); err != nil {
-		return fmt.Errorf("sending open response: %w", err)
+	if address == "" {
+		opened.OpenResponse.Status = pf.TaskNetworkProxyResponse_PORT_NOT_ALLOWED
+		return client.Send(opened)
 	}
 
-	handshakeComplete = true
-	var shardID = open.ShardId.String()
-	proxyConnectionsAcceptedCounter.WithLabelValues(shardID, strconv.Itoa(int(open.TargetPort))).Inc()
-	ops.PublishLog(container.logger, ops.Log_debug, "proxy connection opened", "port", open.TargetPort, "clientIP", open.ClientAddr)
-
-	var grp = errgroup.Group{}
-
-	grp.Go(func() error {
-		if e := copyResponses(streaming, proxyClient, shardID, open.TargetPort, &outboundBytes); isFailure(e) {
-			return fmt.Errorf("copying outbound data: %w", e)
-		}
-		return nil
-	})
-	grp.Go(func() error {
-		if e := copyRequests(streaming, proxyClient, shardID, open.TargetPort, &inboundBytes); isFailure(e) {
-			return fmt.Errorf("copying inbound data: %w", e)
-		}
-		return nil
-	})
+	// Dial the container.
+	var dialer net.Dialer
+	dialed, err := dialer.DialContext(ctx, "tcp", address)
+	var delegate = dialed.(*net.TCPConn)
+	defer delegate.Close()
 
-	err = grp.Wait()
-	var status = "ok"
 	if err != nil {
-		status = "error"
-		ops.PublishLog(container.logger, ops.Log_warn, "proxy connection failed", "port", open.TargetPort, "clientAddr", open.ClientAddr, "error", err)
-	} else {
-		ops.PublishLog(container.logger, ops.Log_debug, "proxy connection closing normally", "port", open.TargetPort, "clientAddr", open.ClientAddr)
+		return fmt.Errorf("failed to connect to container: %w", err)
 	}
-	proxyConnectionsClosedCounter.WithLabelValues(shardID, strconv.Itoa(int(open.TargetPort)), status).Inc()
-	return nil
-}
 
-func (ps *ProxyServer) openConnection(ctx context.Context, open *pf.TaskNetworkProxyRequest_Open) (*pf.TaskNetworkProxyResponse_OpenResponse, *runningContainer, error) {
-	shardStatus, header, err := ps.checkShardStatus(ctx, open)
-	if err != nil {
-		return nil, nil, fmt.Errorf("resolving shard status: %w", err)
-	}
-	var resp = &pf.TaskNetworkProxyResponse_OpenResponse{
-		Status: pf.TaskNetworkProxyResponse_Status(shardStatus),
-		Header: &header,
-	}
-	if resp.Status != pf.TaskNetworkProxyResponse_OK {
-		return resp, nil, nil
-	}
+	// All validations were successful and we dialed the container.
+	_ = client.Send(opened)
```

Comment:
Seems kind of weird to ignore the error here, is this something that can just never error given that the container has already been dialed?

Reply:
It can error but the error isn't useful -- it's only ever an io.EOF placeholder that means "i've been reset", and the _real_ error is only read out by doing Recv().

Note that we ignore it here, but we do check Send error below as an indication that we can cancel ongoing loop work.